modules/javafx.controls/src/test/java/test/javafx/scene/control/SelectionFocusModelMemoryTest.java

```diff
@@ -0,0 +1,298 @@
+        control.setSelectionModel(replacingSm);
+        attemptGC(weakRef, 10);
+        assertNull("selectionModel must be gc'ed", weakRef.get());
+    }
+
+    @Test
+    public void testComboBoxSelectionModel() {
+        ComboBox<String> control = new ComboBox<>(FXCollections.observableArrayList("Apple", "Orange", "Banana"));
+        WeakReference<SelectionModel<?>> weakRef = new WeakReference<>(control.getSelectionModel());
+        SingleSelectionModel<String> replacingSm = ComboBoxShim.get_ComboBoxSelectionModel(control);
+        maybeShowControl(control);
+        control.setSelectionModel(replacingSm);
+        attemptGC(weakRef, 10);
+        assertNull("selectionModel must be gc'ed", weakRef.get());
+    }
+
+    @Test
+    public void testChoiceBoxSelectionModel() {
+        // FIXME
+        // can't formally ignore just one parameter, so backing out if showBeforeReplaceSM
+        // will be fixed as side-effect of skin cleanup in https://bugs.openjdk.java.net/browse/JDK-8087555
+        if (showBeforeReplaceSM) return;
+        ChoiceBox<String> control = new ChoiceBox<>(FXCollections.observableArrayList("Apple", "Orange", "Banana"));
+        WeakReference<SelectionModel<?>> weakRef = new WeakReference<>(control.getSelectionModel());
+        SingleSelectionModel<String> replacingSm = ChoiceBoxShim.get_ChoiceBoxSelectionModel(control);
+        maybeShowControl(control);
+        control.setSelectionModel(replacingSm);
+        attemptGC(weakRef, 10);
+        assertNull("selectionModel must be gc'ed", weakRef.get());
+    }
+
+    private void attemptGC(WeakReference<?> weakRef, int n) {
+        // Attempt gc n times
+        for (int i = 0; i < n; i++) {
+            System.gc();
+            System.runFinalization();
+
+            if (weakRef.get() == null) {
+                break;
+            }
+            try {
+                Thread.sleep(500);
+            } catch (InterruptedException e) {
+               System.err.println("InterruptedException occurred during Thread.sleep()");
+            }
+        }
+    }
+
+    protected void maybeShowControl(Control control) {
+        if (!showBeforeReplaceSM) return;
+        show(control);
+    }
+
+// ------------- parameterized
+
+    @Parameterized.Parameters
+    public static Collection<Object[]> data() {
+        // show the control before replacing the selectionModel
+        Object[][] data = new Object[][] {
+            {false},
+            {true },
+        };
+        return Arrays.asList(data);
+    }
+
+    public SelectionFocusModelMemoryTest(boolean showBeforeReplaceSM) {
+        this.showBeforeReplaceSM = showBeforeReplaceSM;
+    }
+
+//------------------ setup
+
+    private void show(Control node) {
+        if (root == null) {
+            root =  new VBox();
+            scene = new Scene(root);
+            stage = new Stage();
+            stage.setScene(scene);
+        }
+        root.getChildren().add(node);
```

Comment:
Will it be good to add a call to `root.getChildren().removeAll()` before adding the `node` to `root` ?

Reply:
> 
> Will it be good to add a call to `root.getChildren().removeAll()` before adding the `node` to `root` ?

hmm .. don't think that it's necessary to clear out other children: the thing that actually matters is that the control goes active in the scenegraph and has a skin attached. Can you think of any context where any of the tests might get whacky if there are more than one nodes in the root? 

Actually, that's a pattern (*cough, read that's more or less c&p'ed) I use in other tests, where it sometimes is allowed and expected that there are more than one.

Reply:
Agree that It would not cause any problem in test execution.